### PR TITLE
Refactor NebulaDaoBasic,

### PR DESCRIPTION
**Pull request by CorvusYe**

### CODE REVIEW

File: ngbatis-demo/src/test/java/ye/weicheng/ngbatis/demo/AbstractBasicDaoTests.java [hunk 1]
```diff
@@ -65,8 +65,8 @@ public void selectBySelective() {
     }
 
     @Test
-    public void deleteLogicById () {
-        int row = repository.deleteLogicById("取名有点难2");
+    public void deleteWithEdgeById () {
+        int row = repository.deleteWithEdgeById("取名有点难2");
```
Comment (wey-gu) [inline ngbatis-demo/src/test/java/ye/weicheng/ngbatis/demo/AbstractBasicDaoTests.java:69]: 😂

**Review by wey-gu — APPROVED** [review]

File: ngbatis-demo/src/test/java/ye/weicheng/ngbatis/demo/AbstractBasicDaoTests.java (hunk 1 above)
Comment (CorvusYe) [reply ngbatis-demo/src/test/java/ye/weicheng/ngbatis/demo/AbstractBasicDaoTests.java:69]: 之前比较皮，哈哈哈
这次提交改动比较大，有一些公开方法移除掉了。
deleteLogicById 名字跟作用也不符，就也改掉了。

File: src/main/java/org/nebula/contrib/ngbatis/proxy/NebulaDaoBasicExt.java [hunk 2]
```diff
@@ -299,42 +123,37 @@ public static String getCqlTpl() {
...
-        String format = String.format("%s", null);
-        List<String> strings = Arrays.asList("dd", format, "ees");
-        System.out.println( String.join( ",", strings ));
+    public static MethodModel getMethodModel() {
+        StackTraceElement stackTraceElement = Thread.currentThread().getStackTrace()[2];
+        MethodModel methodModel = new MethodModel();
+        String methodName = stackTraceElement.getMethodName();
+        Class<?>  dao = getClassFromStack( stackTraceElement );
+
+        assert dao != null;
```
Comment (Nicole00) [inline src/main/java/org/nebula/contrib/ngbatis/proxy/NebulaDaoBasicExt.java:141]: suggest not use assert, except during the initialization.
how about print log info and throw exception?

Comment (CorvusYe) [reply src/main/java/org/nebula/contrib/ngbatis/proxy/NebulaDaoBasicExt.java:141]: Good idea!

File: src/main/java/org/nebula/contrib/ngbatis/SessionDispatcher.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+//
+// This source code is licensed under Apache 2.0 License.
+
+import org.nebula.contrib.ngbatis.session.LocalSession;
+
+/**
+ * 本地会话调度器接口
+ *
+ * @author yeweicheng
+ * @since 2022-08-26 4:34
```
Comment (wey-gu) [inline src/main/java/org/nebula/contrib/ngbatis/SessionDispatcher.java:13]: 经 @Nicole00 提醒才发现你总是这么熬夜，注意身体呀~~

File: src/main/resources/beetl.properties [hunk 4]
```diff
@@ -0,0 +1,13 @@
+
+# Copyright (c) 2022 All project authors and nebula-contrib. All rights reserved.
+#
+# This source code is licensed under Apache 2.0 License.
+
+FN.ng.valueFmt =org.nebula.contrib.ngbatis.binding.beetl.functions.ValueFmtFn
```
Comment (wey-gu) [inline src/main/resources/beetl.properties:6]: windows/dos new line characters?

File: pom.xml [hunk 5]
```diff
@@ -3,7 +3,7 @@
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
-    <packaging>pom</packaging>
+    <packaging>jar</packaging>
```
Comment (Nicole00) [inline pom.xml:6]: thanks, good catch!

File: src/main/java/org/nebula/contrib/ngbatis/binding/beetl/functions/IdFn.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                }
+            }
+            return fnCall( valueFmtFn, id );
+        } catch (IllegalAccessException e) {
+            log.error( e.getMessage() );
+            return null;
+        }
+    }
+
+}
```
Comment (Nicole00) [inline src/main/java/org/nebula/contrib/ngbatis/binding/beetl/functions/IdFn.java:64]: please add extra empty line. 
image

File: src/main/java/org/nebula/contrib/ngbatis/binding/beetl/functions/TagNameFn.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        if( cm != null ) {
+            return entityType(cm.getClazz());
+        }
+        Class<?> entityType = para.getClass();
+        Table tableAnno = entityType.getAnnotation( Table.class );
+        return tableAnno != null
+                ? tableAnno.name()
+                : StringUtil.xX2x_x(entityType.getSimpleName());
+    }
+}
```
Comment (Nicole00) [inline src/main/java/org/nebula/contrib/ngbatis/binding/beetl/functions/TagNameFn.java:33]: ditto

File: src/main/java/org/nebula/contrib/ngbatis/config/EnvConfig.java [hunk 8]
```diff
@@ -20,6 +22,8 @@
 @Configuration
 public class EnvConfig {
 
+    public static boolean reconnect = true;
```
Comment (Nicole00) [inline src/main/java/org/nebula/contrib/ngbatis/config/EnvConfig.java:25]: is there any code to reset `reconnect`?

**Review by wey-gu — APPROVED** [review]
Thanks!

File: src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+            releaseInnerSession(session);
+        }
+    }
+
+    @Override
+    public LocalSession poll() {
+        LocalSession localSession = null;
+        if( !EnvConfig.reconnect ) {
+            localSession = newLocalSession();
+            localSession.useCount ++;
```
Comment (Nicole00) [inline src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java:76]: if users use the executeWithParameter in mutil threads, how to keep the `poll` function thread-safe?

localSession.useCount++ is not safe here.

Comment (CorvusYe) [reply src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java:76]: Do you mean that different users use the same session to execute different nGQLs

File: src/main/java/org/nebula/contrib/ngbatis/config/EnvConfig.java (hunk 8 above)
Comment (CorvusYe) [reply src/main/java/org/nebula/contrib/ngbatis/config/EnvConfig.java:25]: Not. I want to leave it to the developers to make a decision in their project.

File: src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java (hunk 9 above)
Comment (CorvusYe) [reply src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java:76]: If I add synchronized to the poll method, can it help?

**Comment by CorvusYe** [comment]
It doesn't seem to make much difference.
- Settings in JMeter
image
- With synchronized：
image
- No synchronized：
image

File: src/main/java/org/nebula/contrib/ngbatis/config/EnvConfig.java (hunk 8 above)
Comment (Nicole00) [reply src/main/java/org/nebula/contrib/ngbatis/config/EnvConfig.java:25]: should we provide a setReconnect function?

File: src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java (hunk 9 above)
Comment (Nicole00) [reply src/main/java/org/nebula/contrib/ngbatis/session/IntervalCheckSessionDispatcher.java:76]: Yeah I thank so.

**Comment by Nicole00** [comment]
> It doesn't seem to make much difference.
> 
> * Settings in JMeter
>   image
> * With synchronized：
>   image
> * No synchronized：
>   image

all query execute successfully because the session.execute() is synchronized. I mean  in multi threads in the mean time, poll  should return different session, then there's no such situation that a session is used in different threads.

ps: one session is used in different thread in the mean time is not appropriate

**Review by Nicole00 — APPROVED** [review]

**Comment by CorvusYe** [comment]
> all query execute successfully because the session.execute() is synchronized. I mean in multi threads in the mean time, poll should return different session, then there's no such situation that a session is used in different threads.

> ps: one session is used in different thread in the mean time is not appropriate

What puzzles me is that, when I add sync to `poll` method, query execute failed sometimes.
I don't know if you can reproduce it.

**Pull request merged**